pkg/karmadactl/apply.go

```diff
@@ -0,0 +1,214 @@
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	restclient "k8s.io/client-go/rest"
+	kubectlapply "k8s.io/kubectl/pkg/cmd/apply"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/karmadactl/options"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+// CommandApplyOptions contains the input to the apply command.
+type CommandApplyOptions struct {
+	// global flags
+	options.GlobalCommandOptions
+	// apply flags
+	KubectlApplyFlags *kubectlapply.ApplyFlags
+	Namespace         string
+	AllClusters       bool
+
+	kubectlApplyOptions *kubectlapply.ApplyOptions
+}
+
+var (
+	applyLong = templates.LongDesc(i18n.T(`
+		Apply a configuration to a resource by file name or stdin and propagate them into member clusters.
+		The resource name must be specified. This resource will be created if it doesn't exist yet.
+		To use 'apply', always create the resource initially with either 'apply' or 'create --save-config'.
+
+		JSON and YAML formats are accepted.
+
+		Alpha Disclaimer: the --prune functionality is not yet complete. Do not use unless you are aware of what the current state is. See https://issues.k8s.io/34274.
+		
+		Note: It implements the function of 'kubectl apply' by default. 
+		If you want to propagate them into member clusters, please use 'kubectl apply --all-clusters'.`))
+
+	applyExample = templates.Examples(i18n.T(`
+		# Apply the configuration without propagation into member clusters. It acts as 'kubectl apply'.
+		%[1]s apply -f manifest.yaml
+
+		# Apply resources from a directory and propagate them into all member clusters.
+		%[1]s apply -f dir/ --all-clusters`))
+)
+
+// NewCommandApplyOptions returns an initialized CommandApplyOptions instance
+func NewCommandApplyOptions() *CommandApplyOptions {
+	streams := genericclioptions.IOStreams{In: getIn, Out: getOut, ErrOut: getErr}
+	flags := kubectlapply.NewApplyFlags(nil, streams)
+	return &CommandApplyOptions{
+		KubectlApplyFlags: flags,
+	}
+}
+
+// NewCmdApply creates the `apply` command
+func NewCmdApply(karmadaConfig KarmadaConfig, parentCommand string) *cobra.Command {
+	o := NewCommandApplyOptions()
+	cmd := &cobra.Command{
+		Use:                   "apply (-f FILENAME | -k DIRECTORY)",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Apply a configuration to a resource by file name or stdin and propagate them into member clusters"),
+		Long:                  applyLong,
+		Example:               fmt.Sprintf(applyExample, parentCommand),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(karmadaConfig, cmd, parentCommand, args))
```

Comment:
We can uniformly use `RunE` to throw error that will be dealed at `cmd`, refering #2162

Reply:
We talked it at https://github.com/karmada-io/karmada/pull/886#pullrequestreview-792707940, let @RainbowMango to confirm again.